components/fxa-client/src/send_tab.rs

```diff
@@ -0,0 +1,77 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+pub use crate::commands::send_tab::{SendTabPayload, TabData};
+use crate::{
+    commands::send_tab::{self, EncryptedSendTabPayload, PrivateSendTabKeys, PublicSendTabKeys},
+    errors::*,
+    http_client::GetDeviceResponse,
+    scopes, FirefoxAccount,
+};
+
+impl FirefoxAccount {
+    /// Ensures that the send tab command has been registered for our own device.
+    ///
+    /// **💾 This method alters the persisted account state.**
+    pub fn ensure_send_tab_registered(&mut self) -> Result<()> {
+        let own_keys: PrivateSendTabKeys =
+            match self.state.commands_data.get(send_tab::COMMAND_NAME) {
+                Some(s) => serde_json::from_str(s)?,
+                None => {
+                    let keys = PrivateSendTabKeys::from_random()?;
+                    self.state.commands_data.insert(
+                        send_tab::COMMAND_NAME.to_owned(),
+                        serde_json::to_string(&keys)?,
+                    );
+                    keys
+                }
+            };
+        let public_keys: PublicSendTabKeys = own_keys.into();
+        let kek = self.sync_keys_as_send_tab_kek()?;
+        let command_data: String = public_keys.as_command_data(&kek)?;
+        self.register_command(send_tab::COMMAND_NAME, &command_data)?;
+        Ok(())
+    }
+
+    /// Send a single tab to another device designated by its device ID.
+    pub fn send_tab(&self, target_device_id: &str, title: &str, url: &str) -> Result<()> {
+        let devices = self.get_devices()?;
+        let target = devices
+            .iter()
+            .find(|d| d.id == target_device_id)
+            .ok_or_else(|| ErrorKind::UnknownTargetDevice(target_device_id.to_owned()))?;
+        let payload = SendTabPayload::single_tab(title, url);
+        let kek = self.sync_keys_as_send_tab_kek()?;
+        let command_payload = send_tab::build_send_command(&kek, target, &payload)?;
+        self.invoke_command(send_tab::COMMAND_NAME, target, &command_payload)
+    }
+
+    pub(crate) fn handle_send_tab_command(
+        &self,
+        sender: Option<GetDeviceResponse>,
+        payload: serde_json::Value,
+    ) -> Result<(Option<GetDeviceResponse>, SendTabPayload)> {
+        let send_tab_key: PrivateSendTabKeys =
+            match self.state.commands_data.get(send_tab::COMMAND_NAME) {
+                Some(s) => serde_json::from_str(s)?,
+                None => {
+                    return Err(ErrorKind::IllegalState(
+                        "Cannot find send-tab keys. Has ensure_send_tab been called before?"
+                            .to_string(),
+                    )
+                    .into());
+                }
+            };
+        let encrypted_payload: EncryptedSendTabPayload = serde_json::from_value(payload)?;
+        Ok((sender, encrypted_payload.decrypt(&send_tab_key)?))
+    }
```

Comment:
One API method that might be missing here, is a method to list all available send-tab targets.

IIUC the only way for the calling code to do this right now is to call `get_devices` and then iterate over each device, looking for the ones that have registered the `open-uri` command, remembering to skip itself, and maybe even checking that their `open-uri` command data is encrypted with the correct `kid`.

Can we (and should we?) hide that complexity behind a `get_send_tab_target_devices()` method of some kind, rather than requiring the calling code to know about the internal implementation details of the `open-uri` command?

An advantage of doing it this way is that we may be able to split the available targets into several categories:

* The device advertises send-tab and we're able to send to it.
* The device advertises send-tab, but we can't send to it for some reason (maybe it's command data is encrypted with an old key or something).
   * The app could show these greyed-out or similar
* The device does not support send-tab at all
   * Fennec would be in this list; maybe the app wants to show them greyed out or similar?

@grigoryk would love to get your thoughts on this one.

Reply:
This  got a thumbs-up but doesn't look like we ended up going in that direction; future work, or did it not seem worth it in practice in the end?

Reply:
Sorry I should have commented on that. @grigoryk and I settled on populating a `capabilities` set on every device, so that Android can do the filtering itself (and avoid having to grow a new `getDevice` method for each new command in the future).